content/6_Advanced/Lagrange.mdx

```diff
@@ -27,6 +23,145 @@ adding lambda\*smth
+	frameborder="0"
+/>
+
+Here is where the fact that $f(x)$ is concave comes in. Because the slope is non-increasing, we know that $f(x) - \lambda x$ will first increase, then stay the same, and finally decrease.
+
+Let $v(\lambda)$ be the optimal maximal achievable sum with $\lambda$ penalty and $c(\lambda)$ be the number of subarrays used to achieve $v(\lambda)$ (note that if there are multiple such possibilities, we set $c$ to be the **minimal** number of subarrays to achieve $v$). These values can be calculated in $\mathcal{O}(N)$ time using the dynamic programming approach described above.
+
+When we assign the penalty of $\lambda$, we are trying to find the maximal sum if creating a subarray reduces our sum by $\lambda$. In other words, **we are trying to find the maximum of $f(x) - \lambda x$**.
+
+Without loss of generality, suppose there exists a slope equal to $\lambda$. Given the shape of $f(x) - \lambda x$, we know that $f(x) - \lambda x$ will be maximized at the points where $\lambda$ is equal to the slope of $f(x)$ (these points are red in the graph above). This means that $c(\lambda)$ will be the point at which $\lambda$ is equal to the slope of $f(x)$ (if there are multiple such points, then $c(\lambda)$ will be the leftmost one).
+
+Now we know exactly what $\lambda$ represents: $\lambda$ is the slope and $c(\lambda)$ is the position with slope equal to $\lambda$ (if there are multiple such positions then $c(\lambda)$ is the leftmost one).
+
+We binary search for $\lambda$ and find the highest $\lambda$ such that $c(\lambda) \le K$. Let the optimal value be $\lambda_{\texttt{opt}}$. Then our answer is $v(\lambda_{\texttt{opt}}) + \lambda_{\texttt{opt}} K$. Note that this works even if $c(\lambda_{\texttt{opt}}) \neq K$ since  $c(\lambda_{\texttt{opt}})$ and $K$ will be on the same line with slope $\lambda_{\texttt{opt}}$ in that case.
+
+Because calculating $v(\lambda)$ and $c(\lambda)$ with the dynamic programming solution described above will take $\mathcal{O}(N)$ time, this solution runs in $\mathcal{O}(N\log{\sum A[i]})$ time.
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define ll long long
+
+const int MAX = 3e5 + 5;
+
+int n, k, A[MAX];
+pair<ll, ll> dp[MAX][2];
+
+pair<ll, ll> better(pair<ll, ll> a, pair<ll, ll> b) {
+	if (a.first == b.first) return (a.second < b.second ? a : b);
+
+	return (a.first > b.first ? a : b);
+}
+
+bool solve(ll lmb) {
+	dp[0][0] = {0, 0};
+	dp[0][1] = {-1e18, 0};
+
+	for (int i = 1; i <= n; i++) {
+		dp[i][0] = better(dp[i - 1][0], dp[i - 1][1]);
+
+		dp[i][1] =
+		    better({dp[i - 1][0].first + A[i] - lmb, dp[i - 1][0].second + 1},
+		           {dp[i - 1][1].first + A[i], dp[i - 1][1].second});
+	}
+	return better(dp[n][0], dp[n][1]).second <= k;
+}
+
+int main() {
+	ios_base::sync_with_stdio;
+	cin.tie(0);
+	cin >> n >> k;
+
+	for (int i = 1; i <= n; i++) cin >> A[i];
+
+	ll L = 0, H = 1e15;
+
+	while (L < H) {
+		ll M = (L + H) / 2;
+		solve(M) ? H = M : L = M + 1;
```

Comment:
also names should be in lowercase at least

Reply:
ok i changed it to lo and hi